Restructured Stream your data to an AWS Kinesis Firehose, Azure Event Hub, or GCP Pub/Sub doc (#21369)
newrelic/docs-website: src/content/docs/apis/nerdgraph/examples/nerdgraph-streaming-export.mdx

---
title: "NerdGraph tutorial: Stream your data to an AWS Kinesis Firehose, Azure Event Hub, or GCP Pub/Sub"
tags:
  - APIs
  - NerdGraph
metaDescription: "With the New Relic streaming export feature, you can send your data as it's ingested to New Relic to an AWS Kinesis Firehose, Azure Event Hub or GCP Pub/Sub."
redirects:
freshnessValidatedDate: never
---

With the streaming export feature available through [Data Plus](/docs/accounts/accounts-billing/new-relic-one-pricing-billing/data-ingest-billing/#data-plus), you can send your data to AWS Kinesis Firehose, Azure Event Hub, or GCP Pub/Sub. This guide explains how to create and update streaming rules using [NerdGraph](/docs/apis/nerdgraph/get-started/introduction-new-relic-nerdgraph) and view the existing rules. You can use the [NerdGraph explorer](/docs/apis/nerdgraph/get-started/nerdgraph-explorer) to make these calls.

## What is streaming export? [#definition]

> As data is ingested, our streaming export feature allows you to send the data directly to AWS Kinesis Firehose, Azure Event Hub, or GCP Pub/Sub. You can create custom rules using [NRQL](/docs/query-your-data/nrql-new-relic-query-language/get-started/introduction-nrql-new-relics-query-language) to specify which data should be exported. Additionally, you have the option to compress the data before exporting with our [Export Compression](#compression) feature.

Reviewer: We seem to be repeating the same information over and over again.

* Nested queries are not supported.
* Supports [data types stored in NRDB](/docs/data-apis/understand-data/new-relic-data-types/#timeslice-data), and not [metric timeslice data](/docs/data-apis/understand-data/new-relic-data-types/#timeslice-data).

## Prerequisites

* A [New Relic account](https://newrelic.com/signup), and with that account, you can access your API user key that you need to include with [queries and mutations](/docs/apis/nerdgraph/get-started/introduction-new-relic-nerdgraph/).
* User type and assigned roles can affect your NerdGraph permissions. For more details, see [Factors affecting access](/docs/accounts/accounts-billing/account-structure/factors-affecting-access-features-data/#user-permissions).

## Set up an AWS Kinesis Firehose [#firehose-setup]

> To set up streaming data export to AWS, you must first set up an Amazon Kinesis Firehose. We'll walk you through that procedure in the following steps.

Reviewer: ```suggestion
To set up streaming data export to AWS, you must first set up Amazon Kinesis Firehose. We'll walk you through that procedure in the following steps.
```

<Steps>
  <Step>
### Create a Firehose for streaming export [#create-firehose]

Create a dedicated Firehose to stream your New Relic data to:

1. Go to [Amazon Kinesis Data Firehose](https://aws.amazon.com/firehose/).
2. Create a delivery stream.
3. Name the stream. You'll use this name later when registering the rule.
4. Use <DNT>**Direct PUT or other sources**</DNT> and specify a destination compatible with New Relic's JSON event format (for example, S3, Redshift, or OpenSearch).

</Step>
  <Step>
### Create IAM Firehose write access policy [#create-policy]

1. Go to the IAM console and sign in with your user.
2. In the left navigation, click **Policies**, and then click **Create policy**.
3. Select the Firehose service, and then select `PutRecord` and `PutRecordBatch`.
4. For `Resources`, select the delivery stream, add ARN, and select the region of your stream.
5. Enter your AWS account number, and then enter your desired delivery stream name in the name box.
6. Create the policy.

</Step>
  <Step>
### Create IAM role for granting New Relic write access [#iam-role]
               
To set up the IAM role:
      

> 1. Navigate to the IAM and click on <DNT>**Roles**</DNT>.

Reviewer: ```suggestion
1. Navigate to the IAM and click <DNT>**Roles**</DNT>.
```

2. Create a role for an AWS account, and then select <DNT>**for another AWS account**</DNT>.
3. Enter the New Relic export account ID: `8886xx727xx`.
4. Select <DNT>**Require external ID**</DNT> and enter the [account ID](/docs/accounts/accounts-billing/account-structure/account-id) of the New Relic account you want to export from.
5. Click <DNT>**Permissions**</DNT>, and then select the policy you created above.
6. Add a role name, which will be used during export registration, and provide a description.
7. Create the role.

</Step>
</Steps>
                     
When you're done with these steps, you can set up your export rules using [NerdGraph](/docs/apis/nerdgraph/get-started/introduction-new-relic-nerdgraph/).

<CollapserGroup>
<Collapser 
    id="create-stream-AWS Kinesis"
    title="Create a stream">
    

>     This API allows you to create a streaming export rule to send data to an AWS Kinesis Firehose.

Reviewer: ```suggestion
    This API allows you to create a streaming export rule to send data to AWS Kinesis Firehose.
```

  }
}
```
    **Response**: You receive a response with the rule ID and status. <br />
    **Points to note**:
    When updating, you'll get a message in the message field: “The export rule is being updated and the process may take a few minutes to complete. Please check again later.” It can take up to six minutes to be fully updated.

>     You can check if the rule is updated by calling [streamingRule](view-stream-rule-details) to retrieve the rule. During the period that the rule is under updating, the rule is locked, and no other mutation action can act on the rule. If you are trying to perform another mutation action on the same rule, you will get a message saying, “The export rule is currently being updated by another request, please wait and try again later.” A user can update a rule of any status except a deleted rule.

Reviewer: ```suggestion
    You can check whether the rule is updated by calling [streamingRule](view-stream-rule-details) to retrieve the rule. During the period that the rule is under updating, the rule is locked, and no other mutation action can act on the rule. If you are trying to perform another mutation action on the same rule, you will get a message saying, “The export rule is currently being updated by another request, please wait and try again later.” A user can update a rule of any status except a deleted rule.
```

    message
  }
}
``` 
    **Response**: If you're trying to disable a rule that's in another state, it returns the error message like, "The export rule can't be enabled or disabled due to status not being allowed." You can't disable the rule if the rule is locked due to another mutation being done.

<Callout variant="important">
  You can only disable the rule when it has a status of ENABLED.
</Callout>

  </Collapser>
  <Collapser 
    id="delete-stream"
    title="Delete a stream">
    

>     This API allows you to delete a streaming rule and stop all data export related to that rule.

Reviewer: what happens to already exported data?

Author: Once the streaming rule gets deleted, you can't stream the data. But, already exported data, resides in third party tool.

    #### Input parameters

    <table>
      <thead>
        <tr>
          <th>Parameter</th>
          <th>Data Type</th>
          <th>Is it Required?</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>`id`</td>
          <td>ID</td>
          <td>Yes</td>
          <td>The id of the streaming rule to delete.</td>
        </tr>
      </tbody>
    </table>

    #### Sample request

```graphql
mutation {
  streamingExportDeleteRule(id: RULE_ID) {
    id
  }
}
```
<Callout variant="important">
  Deleting can be performed on a rule of any status unless it's already deleted. Once a rule is deleted, it can't be reactivated again. The rule can be still viewed within the first 24 hours after deletion by calling the [steamingRule](#view-stream-rule-details) API with the rule ID. After 24 hours, the rule won't be searchable anymore through NerdGraph.
</Callout>

  </Collapser>
  
</CollapserGroup>

## Set up an Azure Event Hub [#event-hub-setup]

> To set up streaming data export to Azure, you must first set up an Event Hub. We'll walk you through that procedure in the following steps:

Reviewer: Who is we? We never refer to ourselves in our documentation. Rewrite.

Alternatively, you can follow the Azure guide [here](https://learn.microsoft.com/en-us/azure/event-hubs/event-hubs-create).

<Steps>
  <Step>
### Create an Event Hubs namespace [#create-event-hubs-namespace]

> 1. Navigate to Event Hubs within your Microsoft Azure account.

Reviewer: ```suggestion
1. From your Microsoft Azure account, navigate to Event Hubs.
```

2. Follow the steps to create an Event Hubs namespace. We recommend enabling auto-inflate to ensure you receive all of your data.
3. Ensure public access is enabled, as we will use a Shared Access Policy to securely authenticate with your Event Hub.
4. Once your Event Hubs namespace is deployed, click <DNT>**Go to resource**</DNT>.
  </Step>
  <Step>
### Create an Event Hub [#create-event-hub]

1. In the left column, click <DNT>**Event Hubs**</DNT>.

> 2. Click <DNT>**+Event Hub**</DNT> to create an Event Hub.

Reviewer: ```suggestion
2. To create an Event Hub, click <DNT>**+Event Hub**</DNT>.
```

<CollapserGroup>
<Collapser 
    id="create-stream-Azure Event Hub"
    title="Create a stream">
    
    This API allows you to create a streaming export rule to send data to an Azure Event Hub.

    #### Input parameters
    <table>
      <thead>
        <tr>
          <th>Parameter</th>
          <th>Data Type</th>
          <th>Is it Required?</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>`accountId`</td>
          <td>Integer</td>
          <td>Yes</td>
          <td>Your New Relic [account ID](/docs/accounts/accounts-billing/account-structure/account-id/).</td>
        </tr>
        <tr>
          <td>`ruleParameters.description`</td>
          <td>String</td>
          <td>No</td>
          <td>A brief description of the rule. Use this to provide any additional information or context about the export rule.</td>
        </tr>
        <tr>
          <td>`ruleParameters.name`</td>
          <td>String</td>
          <td>Yes</td>
          <td>The name of the export rule.</td>
        </tr>
        <tr>
          <td>`ruleParameters.nrql`</td>
          <td>String</td>
          <td>Yes</td>

>           <td>A NRQL to select the telemetry data to export.</td>

Reviewer: ```suggestion
          <td>An NRQL to select the telemetry data to export.</td>
```

        <tr>
          <td>`azureParameters.eventHubName`</td>
          <td>String</td>
          <td>Yes</td>
          <td>This parameter specifies the name of the Event Hub where events will be sent and processed.
          For example: `my-event-hub.`</td>
        </tr>
      </tbody>
    </table>

    #### Sample request

```graphql
mutation {
  streamingExportUpdateRule(
    id: RULE_ID,
    ruleParameters: {
      description: "ADD_RULE_DESCRIPTION",
      name: "PROVIDE_RULE_NAME",
      nrql: "YOUR_NRQL_QUERY",
      payloadCompression: DISABLED
    },
    azureParameters: {
      eventHubConnectionString: "YOUR_EVENT_HUB_SAS_CONNECTION_STRING",
      eventHubName: "YOUR_EVENT_HUB_NAME"
    }
  ) {
    id
    status
  }
}
```
    **Response**: You receive a response with the rule ID and status.
  </Collapser>
  <Collapser 
    id="enable-stream-azureeventhub"
    title="Enable a stream">

    Refer [here](#enable-stream) to enable an existing streaming rule, facilitating the continuous export of data for real-time processing and integration with other services.

  </Collapser>
  <Collapser 
    id="disable-stream-azureeventhub"
    title="Disable a stream">

>     Refer [here](#disable-stream) to disable an existing streaming rule, stopping the continuous export of data.

Reviewer: ```suggestion
    To disable an existing streaming rule, stopping the continuous export of data, refer [here](#disable-stream).
```

  </Collapser>
  <Collapser 
    id="delete-stream-azureeventhub"
    title="Delete a stream">
    

>     Refer [here](#delete-stream) to delete a streaming rule and stop all data export related to that rule.

Reviewer: ```suggestion
    To delete a streaming rule and stop all data export related to that rule, refer [here](#delete-stream).
```

  </Collapser>
</CollapserGroup>

## Set up a GCP Pub/Sub [#gcp-pubsub-setup]

> To set up streaming data export to GCP, you must first set up a Pub/Sub. We'll walk you through that procedure in the following steps:

Reviewer: Again use of "we"

<Steps>
  <Step>
### Create a Pub/Sub topic [#create-pubsub-topic]

1. Navigate to the Pub/Sub page within your GCP Console.
2. Click **Create topic** to start the process.
3. Enter a topic ID and click **Create**.
  </Step>
  <Step>
### Set up permissions on Pub/Sub [#setup-permissions-pubsub]

1. On the right column of the created topic, click **More actions**. A list of actions appears.

> 2. Select **View permissions**. A permission tap appears on the right.

Reviewer: ```suggestion
2. Select **View permissions**. 
```

<CollapserGroup>
<Collapser 
    id="create-stream-GCP-Pub-Sub"
    title="Create a stream">
    
    This API allows you to create a streaming export rule to send data to GCP Pub/Sub.

    #### Input parameters
    <table>
      <thead>
        <tr>
          <th>Parameter</th>
          <th>Data Type</th>
          <th>Is it Required?</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>`accountId`</td>
          <td>Integer</td>
          <td>Yes</td>
          <td>Your New Relic [account ID](/docs/accounts/accounts-billing/account-structure/account-id/).</td>
        </tr>
        <tr>
          <td>`ruleParameters.description`</td>
          <td>String</td>
          <td>No</td>
          <td>A brief description of the rule. Use this to provide any additional information or context about the export rule.</td>
        </tr>
        <tr>
          <td>`ruleParameters.name`</td>
          <td>String</td>
          <td>Yes</td>
          <td>The name of the export rule.</td>
        </tr>
        <tr>
          <td>`ruleParameters.nrql`</td>
          <td>String</td>
          <td>Yes</td>

>           <td>A NRQL to select the telemetry data to export.</td>

Reviewer: ```suggestion
          <td>An NRQL to select the telemetry data to export.</td>
```

        </tr>
        <tr>
          <td>`ruleParameters.payloadCompression`</td>
          <td>String</td>
          <td>No</td>
          <td>[payloadCompression](#compression) accepts two values:

>             * `DISABLED`: Bydefault, Payloads will not be compressed before being exported.

Reviewer: ```suggestion
            * `DISABLED`: By default, payloads will not be compressed before being exported.
```

          For example: `pub-sub-topic-id.`</td>
        </tr>
      </tbody>
    </table>

    #### Sample request

```graphql
mutation {
  streamingExportUpdateRule(
    id: RULE_ID,
    ruleParameters: {
      description: "ADD_RULE_DESCRIPTION",
      name: "PROVIDE_RULE_NAME",
      nrql: "YOUR_NRQL_QUERY",
      payloadCompression: DISABLED
    },
    gcpParameters: {
      gcpProjectId: "project-id",
      pubsubTopicId: "pub-sub-topic-id"
    }
  ) {
    id
    status
  }
}
```
    **Response**: You receive a response with the rule ID and status.
  </Collapser>
  <Collapser 
    id="enable-stream-gcppubsub"
    title="Enable a stream">

    Refer [here](#enable-stream) to enable an existing streaming rule, facilitating the continuous export of data for real-time processing and integration with other services.

  </Collapser>
  <Collapser 
    id="disable-stream-gcppubsub"
    title="Disable a stream">

    Refer [here](#disable-stream) to disable an existing streaming rule, stopping the continuous export of data.

  </Collapser>
  <Collapser 
    id="delete-stream-gcppubsub"
    title="Delete a stream">
    

>     Refer [here](#delete-stream) to delete a streaming rule and stop all data export related to that rule.

Reviewer: ```suggestion
    To delete a streaming rule and stop all data export related to that rule, refer [here](#delete-stream).
```

  </Collapser>
</CollapserGroup>

## Understand export compression [#compression]

> You can choose to compress data before exporting it, although this feature is off by default. Compressing can help prevent exceeding your data limit and lower outbound data costs.

Reviewer: ```suggestion
You can choose to compress data before exporting it. This feature is off by default. Compressing can help prevent exceeding your data limit and lower outbound data costs.
```

  </Collapser>
</CollapserGroup>

### Automatic decompression in Azure

If you're exporting data to Azure, it's possible to view decompressed versions of the objects stored in your event hub using a [Stream Analytics Job](https://learn.microsoft.com/en-us/azure/event-hubs/process-data-azure-stream-analytics). To do so, follow these steps:

1. Follow [this guide](https://learn.microsoft.com/en-us/azure/event-hubs/process-data-azure-stream-analytics) up to step 16.

    * On step 13, you may choose to use the same event hub as the output without breaking anything, though we do not recommend this if you intend to proceed to step 17 and start the job, as doing so as not been tested.

> 2. In the left pane of your streaming analytics job, click <DNT>**Inputs**</DNT>, then click on the input you set up.

Reviewer: ```suggestion
2. In the left pane of your streaming analytics job, click <DNT>**Inputs**</DNT>, then click the input you set up.
```
